pkg/diagnostics/http_monitoring.go

```diff
@@ -307,3 +365,93 @@ func (h *httpMetrics) HTTPMiddleware(next http.Handler) http.Handler {
 		h.ServerRequestCompleted(r.Context(), method, path, status, reqContentSize, respSize, elapsed)
 	})
 }
+
+func (h *httpMetrics) initPathNormalization(config *config.PathNormalization) pathNormalizationConfig {
+	if config == nil {
+		return pathNormalizationConfig{
+			enabled: false,
+		}
+	}
+
+	pn := pathNormalizationConfig{
+		enabled: config.Enabled,
+	}
+
+	if !pn.enabled {
+		return pn
+	}
+
+	pn.virtualIngressMux = http.NewServeMux()
+	pn.virtualIngressMux.Handle("/", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if !h.legacy {
+			rw, ok := w.(*diagUtils.PathNormalizationRW)
+			if !ok {
+				log.Errorf("Failed to cast to PathNormalizationRW")
+				return
+			}
+			rw.NormalizedPath = diagUtils.UnmatchedPath
+		}
+	}))
+	for _, pattern := range config.IngressPaths {
+		pn.virtualIngressMux.Handle(pattern, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			rw, ok := w.(*diagUtils.PathNormalizationRW)
+			if !ok {
+				log.Errorf("Failed to cast to PathNormalizationRW")
+				return
+			}
+			rw.NormalizedPath = pattern
+		}))
+	}
+
+	pn.virtualEgressMux = http.NewServeMux()
+	pn.virtualEgressMux.Handle("/", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if !h.legacy {
+			rw, ok := w.(*diagUtils.PathNormalizationRW)
+			if !ok {
+				log.Errorf("Failed to cast to PathNormalizationRW")
+				return
+			}
+			rw.NormalizedPath = diagUtils.UnmatchedPath
+		}
+	}))
+	for _, pattern := range config.EgressPaths {
+		pn.virtualEgressMux.Handle(pattern, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			rw, ok := w.(*diagUtils.PathNormalizationRW)
+			if !ok {
+				log.Errorf("Failed to cast to PathNormalizationRW")
+				return
+			}
+			rw.NormalizedPath = pattern
+		}))
+	}
+	return pn
+}
+
+func (h *httpMetrics) normalizePath(path string, direction int) (string, bool) {
+	if !h.pathNormalization.enabled {
+		return "", false
+	}
+
+	if path == "" {
+		return "", false
+	}
+
+	if !strings.HasPrefix(path, "/") {
+		path = "/" + path
+	}
+
+	req, err := http.NewRequest(http.MethodGet, path, nil)
```

Comment:
The proposal said

> [...] normalizing metrics without relying on regex's, which are known to be CPU-intensive.

However here you are:

1. Creating `http.Request` objects, which allocates a lot of resources
2. Invoking a HTTP server
3. Have the HTTP server do routing, which internally most likely relies on some regular expressions

If you do want to do this in a lightweight fashion, you should look into other options that don't involve all this work. This is a very hot path and it should be optimized.

Some ideas could include using a router only, possibly finding one that uses optimized trees and doesn't involve regexs.

Reply:
Yep- we should be using `http.Request` to instantiate the object :+1: 

We are not invoking any HTTP server here- we are invoking the mux router to hit a resulting handler. This mux doesn't use `regexp` which the reason why it was chosen.

Reply:
Changed the instantiation to use http.Request{...}.

As @filintod mentions we are not relying on regex evaluation here.

Given that we are registering the handlers once at startup and not invoking the HTTP server I think the solution using the mux routing primitives does not add much overhead other than the allocation of the request object.

Reply:
Could you run some benchmarks please?

Reply:
ofc! I will work on that!

Reply:
**Benchmark Results**

All runs were done over this commit https://github.com/dapr/dapr/pull/7746/commits/1fccd567e0cd39e68b64267c14f7f7e49600cd52 (this will be reverted before merge)

_BenchmarkHTTPMiddlewareWithoutPathNormalization_

```
$  go test -benchtime=100x -benchmem -run=^$ -bench ^BenchmarkHTTPMiddlewareWithoutPathNormalization$ github.com/dapr/dapr/pkg/diagnostics
goos: darwin
goarch: arm64
pkg: github.com/dapr/dapr/pkg/diagnostics
BenchmarkHTTPMiddlewareWithoutPathNormalization-14    	     100	   5669265 ns/op	    4865 B/op	      85 allocs/op
PASS
ok  	github.com/dapr/dapr/pkg/diagnostics	0.951s

$ go test -benchtime=500x -benchmem -run=^$ -bench ^BenchmarkHTTPMiddlewareWithoutPathNormalization$ github.com/dapr/dapr/pkg/diagnostics
goos: darwin
goarch: arm64
pkg: github.com/dapr/dapr/pkg/diagnostics
BenchmarkHTTPMiddlewareWithoutPathNormalization-14    	     500	   5676510 ns/op	    4811 B/op	      84 allocs/op
PASS
ok  	github.com/dapr/dapr/pkg/diagnostics	3.162s

$ go test -benchtime=1000x -benchmem -run=^$ -bench ^BenchmarkHTTPMiddlewareWithoutPathNormalization$ github.com/dapr/dapr/pkg/diagnostics
goos: darwin
goarch: arm64
pkg: github.com/dapr/dapr/pkg/diagnostics
BenchmarkHTTPMiddlewareWithoutPathNormalization-14    	    1000	   5653330 ns/op	    4806 B/op	      84 allocs/op
PASS
ok  	github.com/dapr/dapr/pkg/diagnostics	5.963s
```

_BenchmarkHTTPMiddlewareWithPathNormalization_

```
$ go test -benchtime=100x -benchmem -run=^$ -bench ^BenchmarkHTTPMiddlewareWithPathNormalization$ github.com/dapr/dapr/pkg/diagnostics
goos: darwin
goarch: arm64
pkg: github.com/dapr/dapr/pkg/diagnostics
BenchmarkHTTPMiddlewareWithPathNormalization-14    	     100	   5673615 ns/op	    6440 B/op	     114 allocs/op
PASS
ok  	github.com/dapr/dapr/pkg/diagnostics	0.896s

$ go test -benchtime=500x -benchmem -run=^$ -bench ^BenchmarkHTTPMiddlewareWithPathNormalization$ github.com/dapr/dapr/pkg/diagnostics
goos: darwin
goarch: arm64
pkg: github.com/dapr/dapr/pkg/diagnostics
BenchmarkHTTPMiddlewareWithPathNormalization-14    	     500	   5717264 ns/op	    6342 B/op	     112 allocs/op
PASS
ok  	github.com/dapr/dapr/pkg/diagnostics	3.180s

$ go test -benchtime=1000x -benchmem -run=^$ -bench ^BenchmarkHTTPMiddlewareWithPathNormalization$ github.com/dapr/dapr/pkg/diagnostics
goos: darwin
goarch: arm64
pkg: github.com/dapr/dapr/pkg/diagnostics
BenchmarkHTTPMiddlewareWithPathNormalization-14    	    1000	   5652908 ns/op	    6331 B/op	     112 allocs/op
PASS
ok  	github.com/dapr/dapr/pkg/diagnostics	5.976s
```

_BenchmarkHTTPMiddlewareWithPathNormalizationSyncPool_

```
$ go test -benchtime=100x -benchmem -run=^$ -bench ^BenchmarkHTTPMiddlewareWithPathNormalizationSyncPool$ github.com/dapr/dapr/pkg/diagnostics
goos: darwin
goarch: arm64
pkg: github.com/dapr/dapr/pkg/diagnostics
BenchmarkHTTPMiddlewareWithPathNormalizationSyncPool-14    	     100	   5755573 ns/op	    6179 B/op	     112 allocs/op
PASS
ok  	github.com/dapr/dapr/pkg/diagnostics	0.902s

$ go test -benchtime=500x -benchmem -run=^$ -bench ^BenchmarkHTTPMiddlewareWithPathNormalizationSyncPool$ github.com/dapr/dapr/pkg/diagnostics
goos: darwin
goarch: arm64
pkg: github.com/dapr/dapr/pkg/diagnostics
BenchmarkHTTPMiddlewareWithPathNormalizationSyncPool-14    	     500	   5721990 ns/op	    6049 B/op	     111 allocs/op
PASS
ok  	github.com/dapr/dapr/pkg/diagnostics	3.171s

$ go test -benchtime=1000x -benchmem -run=^$ -bench ^BenchmarkHTTPMiddlewareWithPathNormalizationSyncPool$ github.com/dapr/dapr/pkg/diagnostics
goos: darwin
goarch: arm64
pkg: github.com/dapr/dapr/pkg/diagnostics
BenchmarkHTTPMiddlewareWithPathNormalizationSyncPool-14    	    1000	   5691965 ns/op	    6024 B/op	     111 allocs/op
PASS
ok  	github.com/dapr/dapr/pkg/diagnostics	6.022s
```

Reply:
The number of allocations went up by 1/3, which is quite a bit of added pressure to the GC.
Looks like the sync.Pool didn't help much as there was just 1 allocation saved.

Reply:
I am un-convinced that using the full HTTP request stack is particularly efficient. I don't have an exact number to give you as to what's an acceptable number of added allocations, but it'd be great to limit it.

Maybe consider using a third-party router to integrate? And just the router? https://github.com/julienschmidt/httprouter is supposed to be very fast with very low allocations, if the caveats aren't a concern.

Reply:
@filintod noticed something in the test setup that is interesting which is the fact that since all tests were running in low cardinality the allocation increase is not related to the HTTP routing but to the fact that more tags are added in the requests with path normalization since these include path and method keys/values.

An issue was already opened about this: https://github.com/dapr/dapr/issues/6681

Given this, I ran the benchmarks again with high cardinality and the results with and without path normalization are very similar. This validates that the memory allocation increase observed in the first benchmark run was related to the fact that more tags are added and not to the path-matching implementation. 

_BenchmarkHTTPMiddlewareWithoutPathNormalization_

```
$ go test -benchtime=100x -benchmem -run=^$ -bench ^BenchmarkHTTPMiddlewareWithoutPathNormalization$ github.com/dapr/dapr/pkg/diagnostics

goos: darwin
goarch: arm64
pkg: github.com/dapr/dapr/pkg/diagnostics
BenchmarkHTTPMiddlewareWithoutPathNormalization-14    	     100	   5620625 ns/op	    7681 B/op	     126 allocs/op
PASS
ok  	github.com/dapr/dapr/pkg/diagnostics	0.938s

$ go test -benchtime=500x -benchmem -run=^$ -bench ^BenchmarkHTTPMiddlewareWithoutPathNormalization$ github.com/dapr/dapr/pkg/diagnostics
goos: darwin
goarch: arm64
pkg: github.com/dapr/dapr/pkg/diagnostics
BenchmarkHTTPMiddlewareWithoutPathNormalization-14    	     500	   5627427 ns/op	    7802 B/op	     125 allocs/op
PASS
ok  	github.com/dapr/dapr/pkg/diagnostics	3.136s

$ go test -benchtime=1000x -benchmem -run=^$ -bench ^BenchmarkHTTPMiddlewareWithoutPathNormalization$ github.com/dapr/dapr/pkg/diagnostics
goos: darwin
goarch: arm64
pkg: github.com/dapr/dapr/pkg/diagnostics
BenchmarkHTTPMiddlewareWithoutPathNormalization-14    	    1000	   5639435 ns/op	    7804 B/op	     126 allocs/op
PASS
ok  	github.com/dapr/dapr/pkg/diagnostics	5.949s
```

_BenchmarkHTTPMiddlewareWithPathNormalization_

```
$ go test -benchtime=100x -benchmem -run=^$ -bench ^BenchmarkHTTPMiddlewareWithPathNormalization$ github.com/dapr/dapr/pkg/diagnostics
goos: darwin
goarch: arm64
pkg: github.com/dapr/dapr/pkg/diagnostics
BenchmarkHTTPMiddlewareWithPathNormalization-14    	     100	   5641299 ns/op	    7121 B/op	     126 allocs/op
PASS
ok  	github.com/dapr/dapr/pkg/diagnostics	0.876s

$ go test -benchtime=500x -benchmem -run=^$ -bench ^BenchmarkHTTPMiddlewareWithPathNormalization$ github.com/dapr/dapr/pkg/diagnostics
goos: darwin
goarch: arm64
pkg: github.com/dapr/dapr/pkg/diagnostics
BenchmarkHTTPMiddlewareWithPathNormalization-14    	     500	   5667083 ns/op	    7018 B/op	     124 allocs/op
PASS
ok  	github.com/dapr/dapr/pkg/diagnostics	3.156s

$ go test -benchtime=1000x -benchmem -run=^$ -bench ^BenchmarkHTTPMiddlewareWithPathNormalization$ github.com/dapr/dapr/pkg/diagnostics
goos: darwin
goarch: arm64
pkg: github.com/dapr/dapr/pkg/diagnostics
BenchmarkHTTPMiddlewareWithPathNormalization-14    	    1000	   5670418 ns/op	    7007 B/op	     125 allocs/op
PASS
ok  	github.com/dapr/dapr/pkg/diagnostics	5.980s
```

_BenchmarkHTTPMiddlewareWithPathNormalizationSyncPool_

```
$ go test -benchtime=100x -benchmem -run=^$ -bench ^BenchmarkHTTPMiddlewareWithPathNormalizationSyncPool$ github.com/dapr/dapr/pkg/diagnostics

goos: darwin
goarch: arm64
pkg: github.com/dapr/dapr/pkg/diagnostics
BenchmarkHTTPMiddlewareWithPathNormalizationSyncPool-14    	     100	   5673832 ns/op	    6857 B/op	     124 allocs/op
PASS
ok  	github.com/dapr/dapr/pkg/diagnostics	0.878s

$ go test -benchtime=500x -benchmem -run=^$ -bench ^BenchmarkHTTPMiddlewareWithPathNormalizationSyncPool$ github.com/dapr/dapr/pkg/diagnostics

goos: darwin
goarch: arm64
pkg: github.com/dapr/dapr/pkg/diagnostics
BenchmarkHTTPMiddlewareWithPathNormalizationSyncPool-14    	     500           5687737 ns/op	    6714 B/op	     122 allocs/op
PASS
ok  	github.com/dapr/dapr/pkg/diagnostics	3.160s

$ go test -benchtime=1000x -benchmem -run=^$ -bench ^BenchmarkHTTPMiddlewareWithPathNormalizationSyncPool$ github.com/dapr/dapr/pkg/diagnostics

goos: darwin
goarch: arm64
pkg: github.com/dapr/dapr/pkg/diagnostics
BenchmarkHTTPMiddlewareWithPathNormalizationSyncPool-14    	    1000	   5663766 ns/op	    6712 B/op	     123 allocs/op
PASS
ok  	github.com/dapr/dapr/pkg/diagnostics	5.983s
```